ArduCopter/mode_autorotate.cpp

```diff
@@ -111,145 +164,194 @@ void ModeAutorotate::run()
+            float norm_rpm = g2.arot.get_rpm()/g2.arot.get_hs_set_point();
+            if (norm_rpm > HEAD_SPEED_TARGET_RATIO*1.005f  ||  norm_rpm < HEAD_SPEED_TARGET_RATIO*0.995f) {
+                _target_head_speed -= _hs_decay * last_loop_time_s;
             } else {
                 _target_head_speed = HEAD_SPEED_TARGET_RATIO;
             }
-
             // Set target head speed in head speed controller
             g2.arot.set_target_head_speed(_target_head_speed);
-
             // Run airspeed/attitude controller
-            g2.arot.set_dt(G_Dt);
             g2.arot.update_forward_speed_controller();
-
             // Retrieve pitch target
             _pitch_target = g2.arot.get_pitch();
-
             // Update controllers
-            _flags.bad_rpm = g2.arot.update_hs_glide_controller(G_Dt); //run head speed/ collective controller
+            _flags.bad_rpm = g2.arot.update_hs_glide_controller(); //run head speed/ collective controller
+        } else {
+            _pitch_target = 0.0f;
+            g2.arot.update_hover_autorotation_controller(); //run head speed/ collective controller
+            g2.arot.set_collective_minimum_drag(motors->get_coll_mid());
+            g2.arot.set_entry_sink_rate(curr_vel_z);
+            g2.arot.set_entry_alt(g2.arot.get_ground_distance());
+        }
+
+        break;
+    }
+
+    case Autorotation_Phase::SS_GLIDE: {
+        // Steady state glide functions to be run only once
+        if (_flags.ss_glide_initial == true) {
+
+            gcs().send_text(MAV_SEVERITY_INFO, "SS Glide Phase");
 
-            break;
+            // Set following trim low pass cut off frequency
+            g2.arot.set_col_cutoff_freq(g2.arot.get_col_glide_freq());
+
+            // Set desired forward speed target
+            g2.arot.set_desired_fwd_speed();
+
+            // Set target head speed in head speed controller
+            _target_head_speed = HEAD_SPEED_TARGET_RATIO;  //Ensure target hs is set to glide in case hs hasn't reached target for glide
+            g2.arot.set_target_head_speed(_target_head_speed);
+
+            // Prevent running the initial glide functions again
+            _flags.ss_glide_initial = false;
         }
 
-        case Autorotation_Phase::SS_GLIDE:
-        {
-            // Steady state glide functions to be run only once
-            if (_flags.ss_glide_initial == true) {
+        // Run airspeed/attitude controller
+        g2.arot.update_forward_speed_controller();
 
-                #if CONFIG_HAL_BOARD == HAL_BOARD_SITL
-                    gcs().send_text(MAV_SEVERITY_INFO, "SS Glide Phase");
-                #endif
+        //update sink rate derivative
+        g2.arot.calc_sink_d_avg();
 
-                // Set following trim low pass cut off frequency
-                g2.arot.set_col_cutoff_freq(g2.arot.get_col_glide_freq());
+        //update flare altitude estimate
+        g2.arot.update_flare_alt();
```

Comment:
here there are three calls into the lib that don't do anything here. So it could just be one call. `g2.arot.update_glide()` which would then call these three methods internally. 

The key thing is to try and keep all the complexity in one place. My suggestions are based on the assumption that the complected should be down in the lib and not the mode.